### PR TITLE
[feature] Add crates 'relations' and 'relations_procmacro'

### PR DESCRIPTION
**Pull request by woshilapin**
Need [`typed_index_collection`](https://github.com/CanalTP/typed_index_collection) to be released first.

[Here](https://github.com/woshilapin/relations/actions/runs/62250042) is a link to the build on my fork for now.

### CODE REVIEW

**Comment by woshilapin** [comment]
One thought I had. For now, we would need to include both `relations` and `relations_procmacro` in our `Cargo.toml` if we want to use these crates. But maybe we could take inspiration from `serde` where they're doing something like

```toml
serde = { version = "1", features = ["derive"] }
```

So it might be possible to do something like
```toml
relations = { version = "1", features = ["procmacro"] }
```

File: relations_procmacro/tests/01-get-corresponding.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    assert!(cat_indexes.contains(&cat_1_idx));
+    let cat_2_idx = cats.get_idx("cat_id_2").unwrap();
+    assert!(cat_indexes.contains(&cat_2_idx));
+    let cat_3_idx = cats.get_idx("cat_id_3").unwrap();
+    assert!(cat_indexes.contains(&cat_3_idx));
+    let cat_4_idx = cats.get_idx("cat_id_4").unwrap();
+    assert!(cat_indexes.contains(&cat_4_idx));
+
+    let animal_indexes = model.get_corresponding_from_idx(cat_1_idx);
+    assert!(animal_indexes.contains(&animal_idx));
```
Comment (datanel) [inline relations_procmacro/tests/01-get-corresponding.rs]: it would be good to also test `get_corresponding` with a set o Idx

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+//! Modeling the relations between objects.
+
+/// The error type used by the crate.
+pub type Error = failure::Error;
+
+/// The corresponding result type used by the crate.
+pub type Result<T> = std::result::Result<T, Error>;
```
Comment (datanel) [inline src/lib.rs]: same question as `typed_index_collection`
Do we need to expose it ?
Does someone need to use this Result?

I surely missed something but it is already exposed in `src/relations.rs`

Comment (woshilapin) [reply src/lib.rs]: Ah right, good point. Done in 55e339b89012a1bd1cd793a82d9a104b59504e1f.

File: relations_procmacro/tests/01-get-corresponding.rs (hunk 1 above)
Comment (woshilapin) [reply relations_procmacro/tests/01-get-corresponding.rs]: These were the existing tests that I just copied. I just modified them to not depend on `transit_model` (by introducing `Animal`, `Feline` and `Cat`). However, let me see if I can do something about it.

File: relations_procmacro/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "relations_procmacro"
+description = "Procmacro to help create relations between objects"
+version = "1.0.0"
+authors = ["Kisio Digital <team.coretools@kisio.org>", "Guillaume Pinot <texitoi@texitoi.eu>"]
+edition = "2018"
+license = "MIT"
+repository = "https://github.com/CanalTP/relations"
+keywords = ["macro"]
```
[the remark this answers is not in the document]
Comment (woshilapin) [reply relations_procmacro/Cargo.toml]: Done in dfa1842e960115431b5fb75a27e8024da69e5cc5.

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+keywords = ["relation", "collection"]
+
+[workspace]
+members = [
+  "relations_procmacro",
+]
+
+[dependencies]
+derivative = "1"
+failure = "0.1"
```
[the remark this answers is not in the document]
Comment (woshilapin) [reply Cargo.toml]: Maybe another PR?

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "relations"
+description = "Manage relations between objects"
+version = "1.0.0"
+authors = ["Kisio Digital <team.coretools@kisio.org>", "Guillaume Pinot <texitoi@texitoi.eu>"]
+edition = "2018"
+license = "MIT"
+repository = "https://github.com/CanalTP/relations"
```
[the remark this answers is not in the document]
Comment (woshilapin) [reply Cargo.toml]: Good point, done in d24fe1a530119338f4f29e47670679c7aa15f257.

File: .gitignore [hunk 6]
```diff
@@ -8,3 +8,11 @@ Cargo.lock
...
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+
+#Added by cargo
+#
+#already existing elements were commented out
+
+/target
+#Cargo.lock
```
[the remark this answers is not in the document]
Comment (woshilapin) [reply .gitignore]: Actually, I don't understand these lines either. `cargo new` wrote that file for me. Cleansed in 7e6e7889b918ae7fb859ac3533db3b503faa441e.

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (woshilapin) [reply Cargo.toml]: In the end, I did it in 200d6e27bc96a5ef18cedbc9f72c0ddfc78e7ea1.

File: relations_procmacro/tests/01-get-corresponding.rs (hunk 1 above)
Comment (woshilapin) [reply relations_procmacro/tests/01-get-corresponding.rs]: Done in d36ffef887dd88bda03ab3ce0c93d16a5b4ced35. Let me know if this is what you had in mind?

Comment (datanel) [reply relations_procmacro/tests/01-get-corresponding.rs]: no :) I wanted you to test `model.get_corresponding(idxset)`

**Comment by datanel** [comment]
for `relations = { version = "1", features = ["procmacro"] }` you can do that but I don't know if relations can be used without procmacro. 
I would let procmacro required and optional later if someone needs relations without procmacro

File: relations_procmacro/tests/01-get-corresponding.rs (hunk 1 above)
Comment (woshilapin) [reply relations_procmacro/tests/01-get-corresponding.rs]: Done in 26b59e494f7c37f6c7ef72c04c138e1f13da5546.

**Comment by woshilapin** [comment]
> for `relations = { version = "1", features = ["procmacro"] }` you can do that but I don't know if relations can be used without procmacro.
> I would let procmacro required and optional later if someone needs relations without procmacro

Done in 6881a3c0fc6e2e34fd22c2f55591130ed4f78c4e.

**Pull request merged**